internal/services/oidc.go

```diff
@@ -0,0 +1,250 @@
+package services
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/ydataai/authentication-service/internal/clients"
+	"github.com/ydataai/authentication-service/internal/models"
+	"github.com/ydataai/go-core/pkg/common/logging"
+	"golang.org/x/oauth2"
+)
+
+// OIDCService defines the oidc server struct.
+type OIDCService struct {
+	client         *clients.OIDCClient
+	sessionStorage *models.SessionStorage
+	logger         logging.Logger
+}
+
+// NewOIDCService creates a new OIDC Service.
+func NewOIDCService(logger logging.Logger,
+	client *clients.OIDCClient,
+	sessionStorage *models.SessionStorage) *OIDCService {
+	return &OIDCService{
+		client:         client,
+		sessionStorage: sessionStorage,
+		logger:         logger,
+	}
+}
+
+// GetReadyzFunc make sure if oidc provider is ready.
+func (osvc *OIDCService) GetReadyzFunc() func() bool {
+	return osvc.client.ReadyzFunc
+}
+
+// GetOIDCProviderURL gets OIDC provider URL from the OAuth2 configuration.
+func (osvc *OIDCService) GetOIDCProviderURL(w http.ResponseWriter, r *http.Request) string {
+	// Sets a temporary session with state and nonce to validate
+	// when there is callback from the OIDC Provider.
+	session := osvc.sessionStorage.CreateSession()
+
+	return osvc.client.OAuth2Config.AuthCodeURL(
+		session.State,
+		oidc.Nonce(session.Nonce),
+	)
+}
+
+// GetUserInfo gets User Info from the OAuth2 Token.
+func (osvc *OIDCService) GetUserInfo(ctx context.Context, oauth2Token *oauth2.Token) (*oidc.UserInfo, error) {
+	userInfo, err := osvc.client.Provider.UserInfo(ctx, oauth2.StaticTokenSource(oauth2Token))
+	if err != nil {
+		return nil, err
+	}
+
+	return userInfo, nil
+}
+
+// GetValueFromToken gets the nonce from the ID Token.
+func (osvc *OIDCService) GetValueFromToken(value string, t *models.Tokens) (interface{}, error) {
+	var m map[string]interface{}
+	if err := json.Unmarshal(*t.IDTokenClaims, &m); err != nil {
+		return "", err
+	}
+
+	if key, ok := m[value]; ok {
+		return key, nil
+	}
+	return "", nil
+}
+
+// IsFlowSecure ensures the flow is secure and then, returns a JWT token.
+func (osvc *OIDCService) IsFlowSecure(ctx context.Context,
```

Comment:
Let's think in another name for this method, usually a method with `Is` prefix should return a boolean value.